### PR TITLE
Decentralized Distributed PPO

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation and Context

Implementation of Decentralized Distributed PPO, a simple, synchronous method for distributed RL.

Paper: https://arxiv.org/abs/1911.00357

## How Has This Been Tested

Locally -- adding tests via pytest is kinda odd due to not all builds of pytorch supporting distributed mode.

## Types of changes

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: habitat_baselines/config/default.py [hunk 1]
```diff
@@ -62,13 +62,32 @@
...
+_C.RL.DDPPO.num_recurrent_layers = 2
+_C.RL.DDPPO.backbone = "resnet50"
+_C.RL.DDPPO.pretrained_weights = "data/ddppo-models/gibson-2plus-resnet50.pth"
+# Loads pretrained weights
+_C.RL.DDPPO.pretrained = False
+# Loads just the visual encoder backbone weights
+_C.RL.DDPPO.pretrained_encoder = False
+# Whether or not the visual encoder backbone will be trained
+_C.RL.DDPPO.train_encoder = True
+# Whether or not the reset the critic linear layer
```
Comment (Akarshit) [inline habitat_baselines/config/default.py]:
```suggestion
# Whether or not to reset the critic linear layer
```

File: habitat_baselines/config/pointnav/ddppo_pointnav.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    distrib_backend: NCCL
+    # Visual encoder backbone
+    pretrained_weights: data/ddppo-models/gibson-2plus-resnet50.pth
+    # Initialize with pretrained weights
+    pretrained: False
+    # Initialize just the visual encoder backbone with pretrained weights
+    pretrained_encoder: False
+    # Whether or not the visual encoder backbone will be trained.
+    train_encoder: True
+    # Whether or not the reset the critic linear layer
```
Comment (Akarshit) [inline habitat_baselines/config/pointnav/ddppo_pointnav.yaml]:
```suggestion
    # Whether or not to reset the critic linear layer
```

File: habitat_baselines/config/default.py (hunk 1 above)
Comment (erikwijmans) [reply habitat_baselines/config/default.py]: Fixed!  Thank you.

**Comment by mathfac** [comment]
I am getting `run.py: error: unrecognized arguments: --local_rank=0` when I try to run `single_node.sh`.

**Comment by erikwijmans** [comment]
Looks like I hadn't updated that since pytorch changed `--use_env`

File: habitat_baselines/config/default.py [hunk 3]
```diff
@@ -62,13 +62,32 @@
...
 _C.RL.PPO.max_grad_norm = 0.5
 _C.RL.PPO.num_steps = 5
-_C.RL.PPO.hidden_size = 512
 _C.RL.PPO.use_gae = True
 _C.RL.PPO.use_linear_lr_decay = False
 _C.RL.PPO.use_linear_clip_decay = False
 _C.RL.PPO.gamma = 0.99
 _C.RL.PPO.tau = 0.95
 _C.RL.PPO.reward_window_size = 50
+_C.RL.PPO.use_normalized_advantage = True
```
Comment (mathfac) [inline habitat_baselines/config/default.py]: Please, add `use_normalized_advantage=ppo_cfg.use_normalized_advantage,` in `ppo_trainer.py` when `PPO` object is creating.

File: habitat_baselines/common/rollout_storage.py [hunk 4]
```diff
@@ -55,7 +55,7 @@ def __init__(
             self.actions = self.actions.long()
             self.prev_actions = self.prev_actions.long()
 
-        self.masks = torch.ones(num_steps + 1, num_envs, 1)
+        self.masks = torch.zeros(num_steps + 1, num_envs, 1)
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/common/rollout_storage.py]: Yeah, it was/is a bug.

File: habitat_baselines/rl/ddppo/algo/ddp_utils.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def requeue_job():
+    if SLURM_JOBID is None:
+        return
+
+    if not REQUEUE.is_set():
+        return
+
+    distrib.barrier()
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ddppo/algo/ddp_utils.py]: It's needed for the `scontrol requeue`.  That will cancel the job immediately and we want ever process to have cleaned things up before exiting.

File: habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import torch
+import torch.distributed as distrib
+
+from habitat_baselines.rl.ppo import PPO
+
+EPS_PPO = 1e-5
+
+
+def distributed_mean_and_var(values):
+    world_size = distrib.get_world_size()
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ddppo/algo/ddppo.py]: No, that assert will fail often (which is fine/expected)  `distrib.all_reduce(mean)` computes the sum of all the individual means, but we want the mean of the individual means, so we needs to divide by the number of processes.

File: habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    mean = values.mean()
+    distrib.all_reduce(mean)
+    mean /= world_size
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ddppo/algo/ddppo.py]: See above.  It's the `distrib.all_reduce(mean)` that makes us need to divide again.

File: habitat_baselines/rl/ddppo/algo/ddppo_trainer.py [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+from habitat_baselines.rl.ddppo.policy.resnet_policy import (
+    PointNavResNetPolicy,
+)
+from habitat_baselines.rl.ppo.ppo_trainer import PPOTrainer
+
+
+@baseline_registry.register_trainer(name="ddppo")
+class DDPPOTrainer(PPOTrainer):
+    def __init__(self, config=None):
+        interrupted_state = load_interrupted_state()
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ddppo/algo/ddppo_trainer.py]: It happens on a requeue of the job.  i.e. a job will run for a while, SLURM will preempt it, the job will reschedule itself, and it'll eventually get to run again and then it'll pickup this state.

File: habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    sq_diff = (values - mean).pow(2).mean()
+    distrib.all_reduce(sq_diff)
+    var = sq_diff / world_size
+
+    return mean, var
+
+
+class DecentralizedDistributedMixin:
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ddppo/algo/ddppo.py]: It's a mixin so it doesn't have a parent.  This way you could do the same thing to an A2C algo also, `DDA2C(A2C, DecentralizedDistributedMixin)`

File: habitat_baselines/rl/ddppo/algo/ddppo.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            return advantages
+
+        mean, var = distributed_mean_and_var(advantages)
+
+        return (advantages - mean) / (var.sqrt() + EPS_PPO)
+
+    def init_distributed(self, find_unused_params=False):
+        # NB: Used to hide the hooks from the nn.Module,
+        # so they don't show up in the state_dict
+        class Guard(object):
```
[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ddppo/algo/ddppo.py]: Nope, it isn't, I'll remove it.

File: habitat_baselines/rl/ddppo/algo/ddppo_trainer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mathfac) [reply habitat_baselines/rl/ddppo/algo/ddppo_trainer.py]: Will,  it be logical to extend this functionality to resume failed/hang jobs in future PRs?

[the remark this answers is not in the document]
Comment (erikwijmans) [reply habitat_baselines/rl/ddppo/algo/ddppo_trainer.py]: The general logic is, only the file naming logic is specific to this scheme. For failed jobs, I just added a copy statement to the batch script to pickup from a state with a different job ID.

File: habitat_baselines/rl/ddppo/algo/ddppo_trainer.py [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+                        episode_rewards,
+                        episode_counts,
+                    )
+                    pth_time += delta_pth_time
+                    env_time += delta_env_time
+                    count_steps_delta += delta_steps
+
+                    # This is where the preemption of workers happens.  If a
+                    # worker detects it will be a straggler, it preempts itself!
+                    if (step >= ppo_cfg.num_steps / 4) and int(
```
Comment (mathfac) [inline habitat_baselines/rl/ddppo/algo/ddppo_trainer.py]: Thank you for adding the comment. For understanding that would be good to describe motivation behind `(step >= ppo_cfg.num_steps / 4)` and maybe make a constant `0.25` with a reasonable name.

Comment (mathfac) [reply habitat_baselines/rl/ddppo/algo/ddppo_trainer.py]: Maybe we can name it as `SHORT_ROLLOUT_THRESHOLD` and add comment near definition, regarding why short rollouts hurt.

**Pull request merged**